internal/grid/handlers.go

```diff
@@ -0,0 +1,697 @@
+func (r *RemoteErr) Is(other error) bool {
+	if r == nil || other == nil {
+		return r == other
+	}
+	var o RemoteErr
+	if errors.As(other, &o) {
+		return r == &o
+	}
+	return false
+}
+
+// IsRemoteErr returns the value if the error is a RemoteErr.
+func IsRemoteErr(err error) *RemoteErr {
+	var r RemoteErr
+	if errors.As(err, &r) {
+		return &r
+	}
+	return nil
+}
+
+type (
+	// SingleHandlerFn is handlers for one to one requests.
+	// A non-nil error value will be returned as RemoteErr(msg) to client.
+	// No client information or cancellation (deadline) is available.
+	// Include this in payload if needed.
+	// Payload should be recycled with PutByteBuffer if not needed after the call.
+	SingleHandlerFn func(payload []byte) ([]byte, *RemoteErr)
+
+	// StatelessHandlerFn must handle incoming stateless request.
+	// A non-nil error value will be returned as RemoteErr(msg) to client.
+	StatelessHandlerFn func(ctx context.Context, payload []byte, resp chan<- []byte) *RemoteErr
+
+	// StatelessHandler is handlers for one to many requests,
+	// where responses may be dropped.
+	// Stateless requests provide no incoming stream and there is no flow control
+	// on outgoing messages.
+	StatelessHandler struct {
+		Handle StatelessHandlerFn
+		// OutCapacity is the output capacity on the caller.
+		// If <= 0 capacity will be 1.
+		OutCapacity int
+	}
+
+	// StreamHandlerFn must process a request with an optional initial payload.
+	// It must keep consuming from 'in' until it returns.
+	// 'in' and 'out' are independent.
+	// The handler should never close out.
+	// Buffers received from 'in'  can be recycled with PutByteBuffer.
+	// Buffers sent on out can not be referenced once sent.
+	StreamHandlerFn func(ctx context.Context, payload []byte, in <-chan []byte, out chan<- []byte) *RemoteErr
+
+	// StreamHandler handles fully bidirectional streams,
+	// There is flow control in both directions.
+	StreamHandler struct {
+		// Handle an incoming request. Initial payload is sent.
+		// Additional input packets (if any) are streamed to request.
+		// Upstream will block when request channel is full.
+		// Response packets can be sent at any time.
+		// Any non-nil error sent as response means no more responses are sent.
+		Handle StreamHandlerFn
+
+		// Subroute for handler.
+		// Subroute must be static and clients should specify a matching subroute.
+		// Should not be set unless there are different handlers for the same HandlerID.
+		Subroute string
+
+		// OutCapacity is the output capacity. If <= 0 capacity will be 1.
+		OutCapacity int
+
+		// InCapacity is the output capacity.
+		// If == 0 no input is expected
+		InCapacity int
+	}
+)
+
+type subHandlerID [32]byte
+
+func makeSubHandlerID(id HandlerID, subRoute string) subHandlerID {
+	b := subHandlerID(sha256.Sum256([]byte(subRoute)))
+	b[0] = byte(id)
+	b[1] = 0 // Reserved
+	return b
+}
+
+func (s subHandlerID) withHandler(id HandlerID) subHandlerID {
+	s[1] = 0 // Reserved
```

Comment:
In line 204, we set the id to the index zero; not sure if this is a typo or not

Reply:
They are just swapped here. I will make index 0 first, so it is clearer.